### PR TITLE
Fix and improve TR3 tony

### PR DESCRIPTION
**Pull request by TokyoSU**
- Added BOSS effect.
- Removed global structure and variable BossData.
- Simplified the code and removed item->GetFlag, Item->SetFlag and item->TestFlag function which will be removed later to avoid confusion.
Don't merge this PR, it will be redone in another branch to avoid sophia commit.

### CODE REVIEW

File: TombEngine/Objects/Effects/Boss.cpp [hunk 1]
```diff
@@ -195,7 +195,7 @@ namespace TEN::Effects::Boss
 	}
 
 	// NOTE: Actual death occurs when countUntilDeath >= 60.
-	void ExplodeBoss(int itemNumber, ItemInfo& item, int countUntilDeath, const Vector4& color, bool allowExplosion)
+	void ExplodeBoss(int itemNumber, ItemInfo& item, int countUntilDeath, const Vector4& color, bool greenExplosion, bool allowExplosion)
```
Comment (Lwmte) [inline TombEngine/Objects/Effects/Boss.cpp:198]: I am against introducing such specific arguments. What if I want to add `pinkExplosion` or `catsAndDogsExplosion` later? This must be changed.

File: TombEngine/Objects/Effects/Boss.cpp [hunk 2]
```diff
@@ -217,7 +217,7 @@ namespace TEN::Effects::Boss
 					item.Pose.Position.x + (Random::GenerateInt(0, 127) - 64 * 2),
 					(item.Pose.Position.y - CLICK(2)) + (Random::GenerateInt(0, 127) - 64 * 2),
 					item.Pose.Position.z + (Random::GenerateInt(0, 127) - 64 * 2),
-					2, -2, 2, item.RoomNumber);
+					3 - greenExplosion, -2, greenExplosion * 2, item.RoomNumber);
```
Comment (Lwmte) [inline TombEngine/Objects/Effects/Boss.cpp:220]: What the hell is happening? `greenExplosion` is a bool variable!!!

File: TombEngine/Objects/Effects/Boss.cpp [hunk 3]
```diff
@@ -233,7 +233,7 @@ namespace TEN::Effects::Boss
...
```
Comment (Lwmte) [inline TombEngine/Objects/Effects/Boss.cpp:236]: As above!

File: TombEngine/Objects/Effects/Boss.cpp [hunk 4]
```diff
@@ -251,7 +251,7 @@ namespace TEN::Effects::Boss
 				item.Pose.Position.x + (Random::GenerateInt(0, 127) - 64 * 2),
 				(item.Pose.Position.y - CLICK(2)) + (Random::GenerateInt(0, 127) - 64 * 2),
 				item.Pose.Position.z + (Random::GenerateInt(0, 127) - 64 * 2),
-				2, -2, 2, item.RoomNumber);
+				3 - greenExplosion, -2, greenExplosion * 2, item.RoomNumber);
```
Comment (Lwmte) [inline TombEngine/Objects/Effects/Boss.cpp:254]: As above!

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp [hunk 5]
```diff
@@ -25,55 +25,68 @@ namespace TEN::Entities::Creatures::TR3
...
+	enum class TonyFlameType
 	{
-		bool on;
-		Vector3i pos;
-		int fallspeed;
-		int speed;
-		short yRot;
-		short room_number;
-		TonyFlameType type;
+		NoFlame = 0,
```
Comment (Lwmte) [inline TombEngine/Objects/TR3/Entity/TonyBoss.cpp:37]: Why both enum entries are initialized as 0?

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp [hunk 6]
```diff
@@ -91,28 +104,27 @@ namespace TEN::Entities::Creatures::TR3
...
+			fx->flag1 = (short)flame.type;
 			fx->speed = flame.speed;
 			fx->color = Vector4::Zero;
-			fx->flag1 = flame.type;
-			fx->flag2 = (GetRandomControl() & 3) + 1;
-
 			switch (flame.type)
 			{
-			case T_ZAPPDEBRIS:
+			case TonyFlameType::ZAppDebris:
```
Comment (Lwmte) [inline TombEngine/Objects/TR3/Entity/TonyBoss.cpp:112]: What does `ZApp` even mean?

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp [hunk 7]
```diff
@@ -152,16 +164,13 @@ namespace TEN::Entities::Creatures::TR3
 		sptr->nodeNumber = hand;
 		sptr->spriteIndex = Objects[ID_DEFAULT_SPRITES].meshIndex;
 		sptr->scalar = 1;
-		unsigned char size = (GetRandomControl() & 31) + 32;
-		sptr->size = size;
-		sptr->sSize = size;
-		sptr->dSize = size / 4;
+		unsigned char size = (GetRandomControl() & 31) + 64;
```
Comment (Lwmte) [inline TombEngine/Objects/TR3/Entity/TonyBoss.cpp:167]: Convert to `GenerateInt`.

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp [hunk 8]
```diff
@@ -312,43 +318,53 @@ namespace TEN::Entities::Creatures::TR3
...
+	static TonyFlameType GetDebrisType(TonyFlameType type)
+	{
+		switch (type)
+		{
+		case TonyFlameType::ZApp:
+			return TonyFlameType::ZAppDebris;
+		case TonyFlameType::Dropper:
+			return TonyFlameType::DropperDebris;
+		}
+		return TonyFlameType::RockZAppDebris;
```
Comment (Lwmte) [inline TombEngine/Objects/TR3/Entity/TonyBoss.cpp:337]: Move it to `default` case switch above.

**Review by Lwmte — CHANGES_REQUESTED** [review]

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp [hunk 9]
```diff
@@ -25,55 +25,68 @@ namespace TEN::Entities::Creatures::TR3
...
 		TONY_STATE_WAIT,
 		TONY_STATE_RISE,
-		TONY_STATE_FLOAT,
-		TONY_STATE_ZAPP,
-		TONY_STATE_ROCKZAPP,
-		TONY_STATE_BIGBOOM,
+		TONY_STATE_FLY,
+		TONY_STATE_SHOOT_RIGHT_HAND,
+		TONY_STATE_SHOOT_CEILING,
+		TONY_STATE_FLIPMAP, // Cause an shockwave explosion.
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply TombEngine/Objects/TR3/Entity/TonyBoss.cpp:54]: fixed

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp [hunk 10]
```diff
@@ -25,55 +25,68 @@ namespace TEN::Entities::Creatures::TR3
...
 	void InitialiseTony(short itemNumber)
 	{
-		auto* item = &g_Level.Items[itemNumber];
-		CheckForRequiredObjects(*item); // Lizard is not required.
-		item->ItemFlags[3] = 0;
-		item->SetFlagField((int)BossItemFlags::DeathCount, 0);
-		item->SetFlagField((int)BossItemFlags::ExplodeCount, 0);
+		auto& item = g_Level.Items[itemNumber];
+		InitialiseCreature(itemNumber);
+		CheckForRequiredObjects(item); // Lizard is not required.
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply TombEngine/Objects/TR3/Entity/TonyBoss.cpp:85]: just a copy-paste comment, it's removed.

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp (hunk 6 above)
Comment (TokyoSU) [reply TombEngine/Objects/TR3/Entity/TonyBoss.cpp:112]: Fixed and renamed to what they really mean.

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp (hunk 8 above)
Comment (TokyoSU) [reply TombEngine/Objects/TR3/Entity/TonyBoss.cpp:337]: Fixed.

File: TombEngine/Objects/TR3/Entity/TonyBoss.cpp (hunk 5 above)
Comment (TokyoSU) [reply TombEngine/Objects/TR3/Entity/TonyBoss.cpp:37]: Fixed.

File: TombEngine/Objects/Effects/Boss.cpp (hunk 2 above)
Comment (TokyoSU) [reply TombEngine/Objects/Effects/Boss.cpp:220]: Fixed in develop branch.

File: TombEngine/Objects/Effects/Boss.cpp (hunk 4 above)
Comment (TokyoSU) [reply TombEngine/Objects/Effects/Boss.cpp:254]: Fixed in develop branch.

**Pull request closed without merging**